File: M/python/code.py
class HistoricalArray:
    def __init__(self, size) -> None:
        self.size_ = size
        self.data = dict()
        self.current_era_id = 0
        self.data[self.current_era_id] = dict()

    def set(self, index, value) -> None:
        self.data[self.current_era_id][index] = value

    def get(self, index, era_id) -> int:
        if index in self.data[era_id]:
            return self.data[era_id][index]
        else:
            return 0

    def begin_new_era(self, era_id) -> None:
        self.current_era_id = era_id
        self.data[self.current_era_id] = dict()

size = int(input())
q = int(input())
historical_array = HistoricalArray(size)
for i in range(q):
    query = input().split()
    query_type = query[0]
    if query_type == "set":
        historical_array.set(int(query[1]), int(query[2]))
    elif query_type == "begin_new_era":
        historical_array.begin_new_era(int(query[1]))
    else:
        print(historical_array.get(int(query[1]), int(query[2])))

File: M/cpp/code.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;


struct HistoricalArray {
    int size;
    int currentEra;
    map<int, vector<int>> eras;

    HistoricalArray(int n) : size(n), currentEra(0) {
        // eras.emplace(std::make_pair(0, vector<int>(size, 0)));
    	eras[currentEra] = vector<int>(size, 0);
    }
    
    void beginNewEra(int eraId) {
        // eras.emplace(std::make_pair(eraId, vector<int>(size, 0)));
        // currentEra = eraId;
    	currentEra = eraId;
    	eras[currentEra] = vector<int>(size, 0);
    }

    void set(int index, int value) {
        eras[currentEra][index] = value;
    }

    int get(int index, int eraId) {
        return eras[eraId][index];
    }
};

int main(int argc, char const *argv[]) {
    int n;
    cin >> n;
    HistoricalArray arr(n);
    int q;
    cin >> q;
    for (int i = 0; i < q; i++) {
        string queryType;
        cin >> queryType;
        if (queryType == "set") {
            int index, value;
            cin >> index >> value;
            arr.set(index, value);
        } else if (queryType == "begin_new_era") {
            int eraId;
            cin >> eraId;
            arr.beginNewEra(eraId);
        } else if (queryType == "get") {
            int index, eraId;
            cin >> index >> eraId;
            cout << arr.get(index, eraId) << endl;
        }
    }
}
